### PR TITLE
 Use routing_hash_function from index metadata  to determine which class to use to get the shard routing.

### PR DESCRIPTION
**Pull request by matriv**
This commit will replace: 9b33e4d35e000a7ebe0a7038e9fa5d4d90033732

@seut Is it ok that here: https://github.com/crate/elasticsearch/blob/dobe/5.0/core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java#L185
we use the Murmur3?

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java [hunk 1]
```diff
@@ -189,6 +186,378 @@ public void testBWC() {
...
+        IndexMetaData.Builder metaBuilder = IndexMetaData.builder("test")
+                                                         .settings(settings(Version.CURRENT)).numberOfShards(8)
+                                                         .numberOfReplicas(1)
+                                                         .putMapping(new MappingMetaData(IndexMetaData.DEFAULT_MAPPING_TYPE, mappingSource));
+        IndexMetaData metaData = metaBuilder.build();
+
+        Map<String, Integer> termToShard = new TreeMap<>();
+        termToShard.put("sEERfFzPSI", 1);
+        termToShard.put("cNRiIrjzYd", 7);
+        termToShard.put("BgfLBXUyWT", 5);
```
Comment (mfussenegger) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: is it really necessary to have so many values here?

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 2]
```diff
@@ -1311,4 +1315,45 @@ public static int getRoutingFactor(IndexMetaData sourceIndexMetadata, int target
...
+    private static Method extractRoutingHashFunction(IndexMetaData indexMetaData) {
+        String routingHashFunctionClass = null;
+        try {
+            routingHashFunctionClass =
+                getNested(getNested(getMappingMap(indexMetaData), "_meta"), SETTING_ROUTING_HASH_FUNCTION);
+            if (routingHashFunctionClass == null) {
+                return DEFAULT_HASH_FUNCTION;
+            }
+            Class<?> hashFunctionClass = Class.forName(routingHashFunctionClass);
+            return hashFunctionClass.getDeclaredMethod("hash", String.class);
```
Comment (mfussenegger) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Would be nice if we could do this without reflection.

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java (hunk 1 above)
Comment (matriv) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: Just copied the values from another test:
https://github.com/crate/elasticsearch/pull/101/files/fa53167e2cd1cde3ece4ac6ab4dcdff4556f5519#diff-45ab3e6f3ee74c457de57a5e1eda3dc4R78
Maybe I should just keep one value per shard? (0-7)

File: core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java [hunk 3]
```diff
@@ -189,6 +186,107 @@ public void testBWC() {
...
+            String key = entry.getKey();
+            int shard = randomBoolean()
+                        ? OperationRouting.generateShardId(metaData, key, null)
+                        : OperationRouting.generateShardId(metaData, "foobar", key);
+            assertEquals(shard, entry.getValue().intValue());
+        }
+
+        // DJB
+        // @formatter:off
+        builder = XContentFactory.jsonBuilder()
```
Comment (seut) [inline core/src/test/java/org/elasticsearch/cluster/routing/OperationRoutingTests.java]: nothing is formatted specially here, so no need to switch the formatter off

**Pull request closed without merging**